### ISSUE TITLE
Format Support: Trace Event Format

### ISSUE THREAD

**Issue by uber-mahyar2** [jlfwong/speedscope#77]
Chrome has a simplified trace format which is a json list of trace dictionaries.  This is the main format we generate internally at our company, so it would be nice if speedscope could render them.

This test trace file loads in `chrome://tracing` while it gives an error in speedscope.

[test.json.zip](https://github.com/jlfwong/speedscope/files/2147191/test.json.zip)

**Comment by uber-mahyar2** [comment]
Yeah it's the trace format in that doc.  

It looks very similar to this format that currently works: [Profile-20180628T170616.json.zip](https://github.com/jlfwong/speedscope/files/2147331/Profile-20180628T170616.json.zip)

I cut a much larger trace into that smaller one, since it worked in the chrome trace and it would be simpler to understand.  The bigger trace has a similar problem.  It comes from stack snapshots that we globbed together into spans for the chrome tracer, much like this item in the attached Profile.json:

```
  {
    "pid": 1996,
    "tid": 27395,
    "ts": 11937507063,
    "ph": "X",
    "cat": "toplevel",
    "name": "MessagePumpLibevent::OnLibeventNotification",
    "args": {
      "src_file": "../../mojo/edk/system/channel_posix.cc",
      "src_func": "StartOnIOThread"
    },
    "dur": 61,
    "tdur": 62,
    "tts": 178590489
  },
```

So the original data source it comes from are call trees.  When you give the chrome tracer the `"ph": "X"` type spans, chrome seems to expect they are call trees too, because the rendering goes pretty wonky if it isn't.  What chrome is rendering looks like it's a call tree to me: 

<img width="549" alt="screen shot 2018-06-28 at 5 34 42 pm" src="https://user-images.githubusercontent.com/24603300/42067167-b732f2e2-7af9-11e8-9bf8-0f0967ac892d.png">

But yes, it doesn't guarantee they are call trees.

**Comment by jlfwong** [comment]
I’ll commit to implementing this feature myself if a total of $200 of donations is made to any combination of these specific charities:

- [American Civil Liberties Union](https://www.aclu.org/)
- [/dev/color](https://www.devcolor.org/welcome): /dev/color is a non-profit organization whose mission is to empower Black software engineers to help one another grow into industry leaders.
- [Girls Who Code](https://girlswhocode.com/about-us/): Girls Who Code was founded with a single mission: to close the gender gap in technology

In this particular case, the $200 would go towards basic import of "B" and "E" events across multiple processes/threads, but would not include visualization of asynchronous events or instantaneous events.

If you’d like to contribute to that total, please post a screenshot of your donation receipt in this thread*.

Independent of that, I would accept a PR to implement this, regardless of donation amounts.

*: I'm planning on trying this policy on a number of issues. Donations will apply to the total of exactly one issue, so please don't post the same receipt in multiple issue threads.

**Comment by aras-p** [comment]
Would love this too! We have various internal tools that generate profiling data in Chrome Tracing format (mostly just using "X" events really - start time and duration); I've blogged about our usage of it [here](http://aras-p.info/blog/2017/01/23/Chrome-Tracing-as-Profiler-Frontend/) and [here](http://aras-p.info/blog/2017/08/08/Unreasonable-Effectiveness-of-Profilers/) and [here](http://aras-p.info/blog/2019/01/16/time-trace-timeline-flame-chart-profiler-for-Clang/).

I see that you have the "will implement if donations there happen" above, how would I go about doing that? :)

**Comment by jlfwong** [comment]
@aras-p Thanks for the interest! Basically just donate to one of the 3 listed charities and post a receipt. I'm trusting people not to forge receipts :)

You can see an example of work that was successfully funded here: https://github.com/jlfwong/speedscope/issues/60

**Comment by aras-p** [comment]
Ok done!
<img width="472" alt="screen shot 2019-01-17 at 9 59 14 am" src="https://user-images.githubusercontent.com/348087/51303626-9a44b780-1a3e-11e9-8321-44a759b6c488.png">

**Comment by jlfwong** [comment]
@aras-p Thank you so much! I'll try to get start on this on the weekend :)

**Comment by jlfwong** [comment]
I merged #197 which implements basic support, but I haven't published or deployed the change yet because my internet is in a just barely working state at home right now. I'll try to get the deploy running tomorrow.

**Comment by jlfwong** [comment]
@aras-p This is now live on https://www.speedscope.app/ and published as `speedscope` on npm. Can you please give this a try and see if this meets your needs? There's a bunch of ambiguity about the right way of implementing some of this stuff, so I'd be grateful for feedback.

**Comment by aras-p** [comment]
@jlfwong some of the trace profiles I had work beautifully now, thanks!

Some others fail to load, seemingly as soon as the trace has "thread/process metadata" (like thread name) event of `ph`:`M` type. If they were just ignored, that would be better than not supporting the trace at all.

Attached a couple files that don't load right now. The others (produced by other tools that we have that don't emit `M` events) load fine and look great! I'd like a way to see multiple thread timelines at once, but that sounds like a wholly separate feature request :)

[FailingToLoadExamples.zip](https://github.com/jlfwong/speedscope/files/2784443/FailingToLoadExamples.zip)

**Comment by jlfwong** [comment]
>  I'd like a way to see multiple thread timelines at once, but that sounds like a wholly separate feature request :)

Indeed :) This is unfortunately a lot of work. That's tracked in #117. I don't have a bounty set on that because I'm not actually sure how much work it would be in practice.

> Attached a couple files that don't load right now.
Thanks for taking the time to do this! I'll try to take a look sometime this week or next at the profiles that don't load.

**Comment by jlfwong** [comment]
@aras-p Okay! The issue that was preventing `MacPlayer_frontend.json` and `bugreporter_frontend.json` from importing should now be fixed in speedscope@1.4.1. That build is also live on https://www.speedscope.app/. Please let me know if you run into other issues!

**Comment by aras-p** [comment]
Works beautifully now, thanks!
